src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+// Based on what we decide, the implementation of this struct can be mode more
+// C++-like or could stay more C-like.
+struct AkColReadOp {
+  void* chpl_arr;
+  int64_t *startIdx;
+  std::shared_ptr<parquet::ColumnReader> column_reader;
+  bool hasNonFloatNulls;
+  chplEnum_t nullMode;
+  int64_t row_idx;
+  int64_t numElems;
+  int64_t batchSize;
+  bool* where_null_chpl;
+  const parquet::ColumnDescriptor* col_info;
+
+  // Generic read implementation
+  template<typename ArrowType>
+  int64_t read() {
+    using Types = TypeBundle<ArrowType>;
+    using ChplType = typename Types::ChplType;
+    using ReaderType = typename Types::ReaderType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    int64_t num_read = 0;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    // note that floats are the types that natively support nulls. This generic
+    // read() shouldn't do anything with floats
+    if (nullMode == noNulls || nullMode == onlyFloats) {
+      while (reader->HasNext() && row_idx < numElems) {
+        if((numElems - row_idx) < batchSize) // adjust batchSize if needed
+          batchSize = numElems - row_idx;
+        reader->ReadBatch(batchSize, nullptr, nullptr, &chpl_ptr[row_idx],
+                          &values_read);
+        row_idx+=values_read;
+        num_read+=values_read;
+      }
+    }
+    else {
+      int16_t definition_level; // nullable type and only reading single records in batch
+      while (reader->HasNext() && row_idx < numElems) {
+        (void)reader->ReadBatch(1, &definition_level, nullptr,
+                                &chpl_ptr[row_idx], &values_read);
+        // if values_read is 0, that means that it was a null value
+        if(values_read == 0) {
+          where_null_chpl[row_idx] = true;
+        }
+        row_idx++;
+        num_read++;
+      }
+    }
+    return num_read;
+  }
+
+
+  template<> int64_t read<arrow::Int32Type>() {
+    return _readShortIntegral<TypeBundle<arrow::Int32Type>>();
+  }
+
+  template<> int64_t read<arrow::UInt32Type>() {
+    return _readShortIntegral<TypeBundle<arrow::UInt32Type>>();
+  }
+
+  template<> int64_t read<arrow::FloatType>() {
+    using Types = TypeBundle<arrow::FloatType>;
+    using ReaderType = typename Types::ReaderType;
+    using ChplType = typename Types::ChplType;
+    using PqType = typename Types::PqType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    int64_t num_read = 0;
+    if (nullMode == noNulls) {
+      while (reader->HasNext() && row_idx < numElems) {
+        PqType* tmpArr = (PqType*)malloc(batchSize * sizeof(PqType));
```

Comment:
use `new` instead of `malloc`

Reply:
also, this `tmpArr` is never freed. Consider wrapping in a `std::unique_ptr` (and applying that change to the rest of the code)

Reply:
I added removing this allocation to the list of followups.